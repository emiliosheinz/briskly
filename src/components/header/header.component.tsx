import { signIn, signOut, useSession } from 'next-auth/react'
import Image from 'next/image'
import Link from 'next/link'

// TODO emiliosheinz: Create unit tests for Header component
// TODO emiliosheinz: Create Storybook documentation for Header component
export function Header() {
  const { data: sessionData } = useSession()

  const onClick = sessionData ? () => signOut() : () => signIn()
  const label = sessionData ? 'Sign out' : 'Sign in'

  function renderAuthButton() {
    return (
      <button
        className='hover:text-gray-900 whitespace-nowrap text-base font-medium text-primary-900'
        onClick={onClick}
      >
        {label}
      </button>
    )
  }

  return (
    <header className='relative bg-primary-50'>
      <div className='mx-auto max-w-7xl px-4 sm:px-6'>
        <div className='flex items-center justify-between border-b-2 border-primary-200 py-6 md:justify-start md:space-x-10'>
          <div className='flex justify-start lg:w-0 lg:flex-1'>
            <Link href='/'>
              <span className='sr-only'>Briskly</span>
              {/* TODO emiliosheinz: Change image component to use custom component */}
              <Image
                className='h-8 w-auto sm:h-10'
                src='/images/logo.png'
                width={80}
                height={80}
                alt='Briskly logo'
              />
            </Link>
          </div>
          <div className='flex flex-1 items-center justify-end lg:w-0'>
            {renderAuthButton()}
          </div>
        </div>
      </div>
    </header>
  )
}
