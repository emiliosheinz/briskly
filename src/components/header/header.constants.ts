import {
  ClockIcon,
  StarIcon,
  FireIcon,
  HeartIcon,
  PlusCircleIcon,
  UserCircleIcon,
  WrenchIcon,
} from '@heroicons/react/24/outline'

import { routes } from '~/utils/navigation'

const MENU_OPTIONS = [
  {
    label: 'Para Revisar',
    icon: ClockIcon,
    href: routes.toBeReviewed(),
    isAuthRequired: true,
  },
  {
    label: 'Mais Votados',
    icon: StarIcon,
    href: '#',
    isAuthRequired: false,
  },
  {
    label: 'Mais Recentes',
    icon: FireIcon,
    href: '#',
    isAuthRequired: false,
  },
  {
    label: 'Para Você',
    icon: HeartIcon,
    href: '#',
    isAuthRequired: true,
  },
  {
    label: 'Criar Deck',
    icon: PlusCircleIcon,
    href: routes.createNewDeck(),
    isAuthRequired: false,
  },
  {
    label: 'Meu Perfil',
    icon: UserCircleIcon,
    href: '#',
    isAuthRequired: true,
  },
  {
    label: 'Configurações',
    icon: WrenchIcon,
    href: '#',
    isAuthRequired: true,
  },
]

export const MOBILE_MENU_OPTIONS = [...MENU_OPTIONS]

export const DESKTOP_MENU_OPTIONS = {
  PRIMARY: MENU_OPTIONS.filter(({ isAuthRequired }) => !isAuthRequired),
  SECONDARY: MENU_OPTIONS.filter(({ isAuthRequired }) => isAuthRequired),
}
