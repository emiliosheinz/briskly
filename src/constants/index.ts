import { Visibility } from '@prisma/client'

import type { Option } from '~/components/radio-group'

export const MAX_TOPICS_PER_DECK = 5

export const DECK_VISIBILITY_OPTIONS: Array<Option<Visibility>> = [
  {
    name: 'Público',
    value: Visibility.Public,
    description: 'Todas as pessoas com acesso a este site',
  },
  {
    name: 'Privado',
    value: Visibility.Private,
    description: 'Apenas você e pessoas com acesso a sua conta',
  },
  {
    name: 'Apenas com link',
    value: Visibility.WithLink,
    description: 'Todos com acesso direto ao link do Deck',
  },
]

export const STUDY_SESSION_BOXES = [
  { reviewGapInHours: 24 },
  { reviewGapInHours: 24 * 3 },
  { reviewGapInHours: 24 * 7 },
  { reviewGapInHours: 24 * 15 },
  { reviewGapInHours: 24 * 20 },
]
