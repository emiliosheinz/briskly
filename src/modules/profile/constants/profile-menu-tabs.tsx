import dynamic from 'next/dynamic'

const DecksToBeReviewed = dynamic(() =>
  import('~/components/deck-card-list').then(
    module => module.DecksToBeReviewed,
  ),
)
const UserDecks = dynamic(() =>
  import('../components/user-decks.component').then(module => module.UserDecks),
)

const UserFavoriteDecks = dynamic(() =>
  import('~/components/deck-card-list').then(
    module => module.UserFavoriteDecks,
  ),
)

const DecksWithStudySession = dynamic(() =>
  import('../components/decks-with-study-session.component').then(
    module => module.DecksWithStudySession,
  ),
)

export const profileMenuTabs = [
  {
    name: 'Decks',
    content: UserDecks,
    isProfileOwnerOnly: false,
  },
  {
    name: 'Meus Estudos',
    content: DecksWithStudySession,
    isProfileOwnerOnly: true,
  },
  {
    name: 'Para Revisar',
    content: DecksToBeReviewed,
    isProfileOwnerOnly: true,
  },
  {
    name: 'Meus Favoritos',
    content: UserFavoriteDecks,
    isProfileOwnerOnly: true,
  },
]
