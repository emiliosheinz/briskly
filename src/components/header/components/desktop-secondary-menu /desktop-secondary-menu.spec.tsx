import { fireEvent, render, screen } from '~/utils/tests/react-testing-library'
import { withNextAuth } from '~/utils/tests/with-next-auth'

import { DESKTOP_MENU_OPTIONS } from '../../header.constants'
import { DesktopSecondaryMenu } from './desktop-secondary-menu.component'

describe('Header DesktopSecondaryMenu Component', () => {
  it('should be correctly rendered', () => {
    const component = render(<DesktopSecondaryMenu />)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should only be visible on Desktop', () => {
    render(<DesktopSecondaryMenu />)

    expect(screen.getByTestId('desktop-secondary-menu')).toHaveClass('hidden')
    expect(screen.getByTestId('desktop-secondary-menu')).toHaveClass('md:block')
  })

  it('should initially render only the popover button', () => {
    render(<DesktopSecondaryMenu />)

    expect(screen.getByTestId('popover-button')).toBeInTheDocument()
    expect(screen.queryByTestId('popover-panel')).not.toBeInTheDocument()
  })

  it('should render popover panel when popover button is clicked', () => {
    render(withNextAuth(<DesktopSecondaryMenu />, 'authenticated'))

    fireEvent.click(screen.getByTestId('popover-button'))

    expect(screen.queryByTestId('popover-panel')).toBeInTheDocument()
  })

  it('should render all DESKTOP_MENU_OPTIONS.SECONDARY menus inside popover panel', () => {
    render(withNextAuth(<DesktopSecondaryMenu />, 'authenticated'))

    fireEvent.click(screen.getByTestId('popover-button'))
    const panel = screen.queryByTestId('popover-panel')

    for (const { label } of DESKTOP_MENU_OPTIONS.SECONDARY) {
      expect(panel?.innerHTML).toContain(label)
    }
  })
})
