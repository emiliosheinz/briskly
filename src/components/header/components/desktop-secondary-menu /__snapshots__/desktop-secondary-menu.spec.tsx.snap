// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header DesktopSecondaryMenu Component should be correctly rendered 1`] = `
<DocumentFragment>
  <div
    class="relative hidden md:block"
    data-headlessui-state=""
    data-testid="desktop-secondary-menu"
  >
    <button
      aria-expanded="false"
      class="inline-flex items-center justify-center bg-primary-50 p-1 text-primary-900"
      data-headlessui-state=""
      data-testid="popover-button"
      id="headlessui-popover-button-:r0:"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="h-8 w-8 text-primary-900"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.982 18.725A7.488 7.488 0 0012 15.75a7.488 7.488 0 00-5.982 2.975m11.963 0a9 9 0 10-11.963 0m11.963 0A8.966 8.966 0 0112 21a8.966 8.966 0 01-5.982-2.275M15 9.75a3 3 0 11-6 0 3 3 0 016 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
